CMake updates: add options + use BUILD_SHARED_LIBS + export targets + win32 dll

madebr opened the pull request:
Hello, we are currently reviewing a request for adding tinyspline in conan-center-index at https://github.com/conan-io/conan-center-index/pull/3149
conan-center-index is a central repo for recipes for conan, which is a c/c++ package manager.
Long story short, this will allow c/c++ developers to use  tinyspline with a simple `conan install tinyspline/x.y`

While reviewing, we needed to apply some changes to the cmake build script.
In order for you to benefit from these, I have applied them to the current master.
I have separated the changes as much as possible in patches.

After the patches, the following is possible:
- optionally build examples/tests/docs
  This is useful for us, packages, as we won't add the examples/tests to the packages
- optionally enable `-Werror`or `/WX`.
  This makes the build more robust. Granted, warnings should be avoided as much as possible, but a new compiler version will introduce new warnings which will break future builds. This change allows you, the developer, to fail on errors, but will allow users to ignore these warnings.
- Don't build shared and static libraries at the same time.
  CMake has a [`BUILD_SHARED_LIBS`](https://cmake.org/cmake/help/latest/variable/BUILD_SHARED_LIBS.html) option that will change the behavior of `add_library`.
  It's uncommon for a dependency of a library to require a shared and a static library at the same time.
- Avoid a OpenGL warning. See https://cmake.org/cmake/help/latest/module/FindOpenGL.html#linux-specific.
- Link the system libraries to the `tinyspline` target immediately. By making use of targets, the `TINYSPLINE_C_STATIC_LIBRARIES` cmake variable should not be added to the library list of dependencies/examples.
- The ctest change allows running `make test` in the root directory
- By tagging the symbols in the headers with `__declspec(dllexport)` on a shared build on Windows, it is possible to create a tinyspline dll + import library.
- The python bump on travis is required, because yesterday cpython 3.9 was released.

msteinbeck commented:
Hi @madebr,

thank you for considering adding TinySpline to conan. I've already heard about conan but never used it on my own. I'll review your changes immediately. I guess I should prepare a new release (create a tag) beforehand.

> optionally build examples/tests/docs
This is useful for us, packages, as we won't add the examples/tests to the packages.

You can also create the CMake project from the CMakeLists file in the ´src´ directory. This builds only the library.

> optionally enable -Werroror /WX.
This makes the build more robust. Granted, warnings should be avoided as much as possible, but a new compiler version will introduce new warnings which will break future builds. This change allows you, the developer, to fail on errors, but will allow users to ignore these warnings.

I see your point.

> Don't build shared and static libraries at the same time.
CMake has a BUILD_SHARED_LIBS option that will change the behavior of add_library.
It's uncommon for a dependency of a library to require a shared and a static library at the same time.

I thought it would be convenient to have both, the static and the shared library at the same time. However, if this change is necessary, I won't fight for keeping this "feature".

> Avoid a OpenGL warning. See https://cmake.org/cmake/help/latest/module/FindOpenGL.html#linux-specific.

Nice. Thanks.

> The python bump on travis is required, because yesterday cpython 3.9 was released.

How time flies...

msteinbeck commented on CMakeLists.txt:
```diff
@@ -4,8 +4,24 @@ project(tinyspline)
 set_property(GLOBAL PROPERTY USE_FOLDERS ON)
 
+# Detect tinyspline being built as main project
+if(CMAKE_SOURCE_DIR STREQUAL "${tinysplice_SOURCE_DIR}")
```
tinysplice?

madebr replied:
Is it ok if I push a commit with the fixes or do you prefer a force push with the fixed commits?

msteinbeck replied:
Whatever is the easiest for you.

msteinbeck commented on CMakeLists.txt:
```diff
@@ -4,8 +4,24 @@ project(tinyspline)
+    set(TINYSPLINE_MAIN_PROJECT ON)
+endif()
+
+option(TINYSPLINE_BUILD_EXAMPLES "Build TinySpline tests." ON)
```
tests -> examples

msteinbeck commented on CMakeLists.txt:
```diff
@@ -4,8 +4,24 @@ project(tinyspline)
+
+option(TINYSPLINE_BUILD_EXAMPLES "Build TinySpline tests." ON)
+option(TINYSPLINE_BUILD_TESTS "Build TinySpline tests." ON)
+option(TINYSPLINE_BUILD_DOCS "Build TinySpline docs." ON)
```
docs -> documentation? (I do not insist)

msteinbeck commented on examples/c/CMakeLists.txt:
```diff
@@ -17,6 +17,7 @@ endif()
 ###############################################################################
 ### Setup OpenGL.
 ###############################################################################
+set(OpenGL_GL_PREFERENCE "GLVND")
```
Please reference https://cmake.org/cmake/help/latest/module/FindOpenGL.html#linux-specific.

msteinbeck commented on src/CMakeLists.txt:
```diff
@@ -505,14 +510,18 @@ if(CMAKE_CXX_COMPILER_ID MATCHES "GNU")
-	set(TINYSPLINE_LIBRARY_C_FLAGS "${TINYSPLINE_LIBRARY_C_FLAGS} -Werror")
+	if(TINYSPLINE_MAIN_PROJECT)
+	    set(TINYSPLINE_LIBRARY_C_FLAGS "${TINYSPLINE_LIBRARY_C_FLAGS} -Werror")
+    endif()
```
spaces -> tabs

msteinbeck commented on src/CMakeLists.txt:
```diff
@@ -541,14 +550,18 @@ elseif(CMAKE_CXX_COMPILER_ID MATCHES "Clang")
-	set(TINYSPLINE_LIBRARY_C_FLAGS "${TINYSPLINE_LIBRARY_C_FLAGS} -Werror")
+	if(TINYSPLINE_MAIN_PROJECT)
+	    set(TINYSPLINE_LIBRARY_C_FLAGS "${TINYSPLINE_LIBRARY_C_FLAGS} -Werror")
+    endif()
```
spaces -> tabs

msteinbeck commented on src/CMakeLists.txt:
```diff
@@ -541,14 +550,18 @@ elseif(CMAKE_CXX_COMPILER_ID MATCHES "Clang")
-	set(TINYSPLINE_LIBRARY_CXX_FLAGS "${TINYSPLINE_LIBRARY_CXX_FLAGS} -Werror")
+	if(TINYSPLINE_MAIN_PROJECT)
+	    set(TINYSPLINE_LIBRARY_CXX_FLAGS "${TINYSPLINE_LIBRARY_CXX_FLAGS} -Werror")
+    endif()
```
spaces -> tabs

msteinbeck commented on src/CMakeLists.txt:
```diff
@@ -567,11 +580,15 @@ elseif(CMAKE_CXX_COMPILER_ID MATCHES "MSVC")
-	set(TINYSPLINE_LIBRARY_C_FLAGS "${TINYSPLINE_LIBRARY_C_FLAGS} /WX")
+	if(TINYSPLINE_MAIN_PROJECT)
+	    set(TINYSPLINE_LIBRARY_C_FLAGS "${TINYSPLINE_LIBRARY_C_FLAGS} /WX")
+    endif()
```
spaces -> tabs

msteinbeck commented on src/CMakeLists.txt:
```diff
@@ -693,28 +710,27 @@ list(APPEND TINYSPLINE_CXX_SOURCE_FILES
-set_target_properties(tinyspline_shared PROPERTIES
+add_library(tinyspline ${TINYSPLINE_C_SOURCE_FILES})
+if(NOT BUILD_SHARED_LIBS)
+    target_compile_definitions(tinyspline PUBLIC TINYSPLINE_STATIC)
```
I don't understand what the purpose of this line is. Is it required by ´target_link_libraries´ in the examples? What is `TINYSPLINE_STATIC`

msteinbeck replied:
spaces -> tabs

msteinbeck commented on src/CMakeLists.txt:
```diff
@@ -693,28 +710,27 @@ list(APPEND TINYSPLINE_CXX_SOURCE_FILES
-	PUBLIC ${TINYSPLINE_C_INCLUDE_DIR})
+	RUNTIME DESTINATION ${TINYSPLINE_INSTALL_BINARY_DIR})
+install(EXPORT tinyspline
+    DESTINATION "${TINYSPLINE_INSTALL_LIBRARY_DIR}/${TINYSPLINE_INSTALL_CMAKE_CONFIG_DIR}/tinyspline"
+        NAMESPACE "tinyspline::"
+    FILE tinyspline-targets.cmake)
```
spaces -> tabs

msteinbeck commented on src/CMakeLists.txt:
```diff
@@ -693,28 +710,27 @@ list(APPEND TINYSPLINE_CXX_SOURCE_FILES
+    FILE tinyspline-targets.cmake)
+target_include_directories(tinyspline
+	PUBLIC
+	    $<BUILD_INTERFACE:${TINYSPLINE_C_INCLUDE_DIR}>
+	    $<INSTALL_INTERFACE:${TINYSPLINE_INSTALL_INCLUDE_DIR}>)
```
spaces -> tabs

madebr commented on src/tinyspline.h:
```diff
@@ -5,6 +5,22 @@
+# define TINYSPLINE_SHARED_IMPORT
+#endif
+
+#ifdef TINYSPLINE_STATIC
```
@msteinbeck 
This is the reason for `TINYSPLINE_STATIC`

When building a shared library on Windows using MSVC, you need to explicitly tag the symbols that you want to export. When building a static library, you generally don't want to export the symbols and just keep default visibility.

To compare with gcc, MSVC has always `-fvisibilty=hidden` enabled.

There is a similar change in `src/tinysplinecxx.h`.

msteinbeck replied:
Thanks for clarification.

madebr replied:
There is an albernative to my approach:
1. generate an export header using https://cmake.org/cmake/help/latest/module/GenerateExportHeader.html for each library
2. use these defines in the headers

But this will generate 2 extra headers (one for tinyspline and one for tinysplinecxx) which I felt was too many for a small library.

msteinbeck replied:
> But this will generate 2 extra headers (one for tinyspline and one for tinysplinecxx) which I felt was too many for a small library.

I agree.

msteinbeck commented on src/CMakeLists.txt:
```diff
@@ -693,28 +710,27 @@ list(APPEND TINYSPLINE_CXX_SOURCE_FILES
+	PUBLIC
+	    $<BUILD_INTERFACE:${TINYSPLINE_C_INCLUDE_DIR}>
+	    $<INSTALL_INTERFACE:${TINYSPLINE_INSTALL_INCLUDE_DIR}>)
+target_link_libraries(tinyspline ${TINYSPLINE_C_STATIC_LIBRARIES})
```
Is this call necessary? The shared library already links the required libraries (e.g, https://github.com/msteinbeck/tinyspline/blob/master/src/CMakeLists.txt#L505). Or do we need to remove:

```
set(CMAKE_SHARED_LINKER_FLAGS "${CMAKE_SHARED_LINKER_FLAGS} -lm")
```

and the like?

madebr replied:
Good catch, this should have `PRIVATE` added!

This is cmake target *magic* at work.
See https://cmake.org/cmake/help/latest/command/target_link_libraries.html#libraries-for-a-target-and-or-its-dependents

`PRIVATE` for a static library means:
- a static library is not linked, only archived so the linked libraries cannot be used,
- the dependencies of a static libraries (e.g. an executable) will link to the libraries added here.

`PRIVATE` for a shared library means:
- the linked libraries are used when linking the shared library (so/dll/dylib)
- when a dependency links to the shared library, no extra libraries are added. It is private.

madebr replied:
By using `target_link_libraries`, the targets can be safely exported.

When you execute `make install`, a `tinyspline-targets.cmake` is installed as well.

msteinbeck commented on src/CMakeLists.txt:
```diff
@@ -739,35 +756,35 @@ install(FILES ${CMAKE_BINARY_DIR}/src/tinyspline.pc
+			PUBLIC
+			    $<BUILD_INTERFACE:${TINYSPLINE_CXX_INCLUDE_DIR}>
+			    $<INSTALL_INTERFACE:${TINYSPLICE_INSTALL_INCLUDE_DIR}>)
+    target_link_libraries(tinysplinecxx PRIVATE ${TINYSPLINE_CXX_STATIC_LIBRARIES})
```
See comment: `target_link_libraries(tinyspline ${TINYSPLINE_C_STATIC_LIBRARIES})`.

Also, why are you using `PRIVATE` here but not in `target_link_libraries(tinyspline ${TINYSPLINE_C_STATIC_LIBRARIES})`?

madebr replied:
See https://github.com/msteinbeck/tinyspline/pull/167#discussion_r502619328

msteinbeck commented on src/CMakeLists.txt:
```diff
@@ -739,35 +756,35 @@ install(FILES ${CMAKE_BINARY_DIR}/src/tinyspline.pc
+        FILE tinysplinecxx-targets.cmake)
+	target_include_directories(tinysplinecxx
+			PUBLIC
+			    $<BUILD_INTERFACE:${TINYSPLINE_CXX_INCLUDE_DIR}>
+			    $<INSTALL_INTERFACE:${TINYSPLICE_INSTALL_INCLUDE_DIR}>)
```
spaces -> tabs

msteinbeck commented on src/CMakeLists.txt:
```diff
@@ -739,35 +756,35 @@ install(FILES ${CMAKE_BINARY_DIR}/src/tinyspline.pc
+	add_library(tinysplinecxx ${TINYSPLINE_CXX_SOURCE_FILES})
+    if(NOT BUILD_SHARED_LIBS)
+        target_compile_definitions(tinysplinecxx PUBLIC TINYSPLINECXX_STATIC)
+    endif()
```
spaces -> tabs

msteinbeck commented on src/CMakeLists.txt:
```diff
@@ -739,35 +756,35 @@ install(FILES ${CMAKE_BINARY_DIR}/src/tinyspline.pc
-	target_include_directories(tinysplinecxx_static
-		PUBLIC ${TINYSPLINE_CXX_INCLUDE_DIR})
+		RUNTIME DESTINATION ${TINYSPLINE_INSTALL_BINARY_DIR})
+    install(EXPORT tinysplinecxx
+        DESTINATION "${TINYSPLINE_INSTALL_LIBRARY_DIR}/${TINYSPLINE_INSTALL_CMAKE_CONFIG_DIR}/tinysplinecxx"
+        NAMESPACE "tinysplinecxx::"
+        FILE tinysplinecxx-targets.cmake)
```
spaces -> tabs

madebr commented:
> thank you for considering adding TinySpline to conan. I've already heard about conan but never used it on my own. I'll review your changes immediately. I guess I should prepare a new release (create a tag) beforehand.

It's @SpaceIm who was the original contributor of the recipe.
He already did a mini-review at https://github.com/conan-io/conan-center-index/pull/3149#discussion_r502388771

> > optionally build examples/tests/docs
> > This is useful for us, packages, as we won't add the examples/tests to the packages.
> 
> You can also create the CMake project from the CMakeLists file in the ´src´ directory. This builds only the library.

Indeed, that would be a possibility.
But we felt that the `src` subdirectory might change.

> > optionally enable -Werroror /WX.
> > This makes the build more robust. Granted, warnings should be avoided as much as possible, but a new compiler version will introduce new warnings which will break future builds. This change allows you, the developer, to fail on errors, but will allow users to ignore these warnings.
> 
> I see your point.

Another possibility would be to add a option to enable `-Werror/-WX`.
Something like the following:
```cmake
option(TINYSPLINE_ENABLE_ERRORS "Enable -Werror/-WX" ON)
```
This allows overriding, even when building as a subproject and does not require a compare in the main cmake script.

> > Don't build shared and static libraries at the same time.
> > CMake has a BUILD_SHARED_LIBS option that will change the behavior of add_library.
> > It's uncommon for a dependency of a library to require a shared and a static library at the same time.
> 
> I thought it would be convenient to have both, the static and the shared library at the same time. However, if this change is necessary, I won't fight for keeping this "feature".

See https://github.com/conan-io/conan-center-index/pull/3149#discussion_r502388771
This pr will rename the libraries on MSVC (no lib prefix).
This will make allow linking to `tinyspline` on msvc using e.g. `link ... tinyspline.lib` and using `gcc ... -ltinyspline`. Note how the `lib` prefix is not needed on MSVC because MSVC does not support the `-l...` syntax.

> > Avoid a OpenGL warning. See https://cmake.org/cmake/help/latest/module/FindOpenGL.html#linux-specific.
> 
> Nice. Thanks.

I will add the comment, as you requested.

msteinbeck commented on src/pkg/tinyspline-config.cmake.in:
```diff
@@ -4,10 +4,8 @@ set_and_check(TINYSPLINE_LIBRARY_DIRS "@PACKAGE_TINYSPLINE_INSTALL_LIBRARY_DIR@"
+include("${CMAKE_CURRENT_LIST_DIR}/tinyspline-targets.cmake")
+set(TINYSPLINE_LIBRARIES tinyspline::tinyspline)
 
 mark_as_advanced(
```
Is `mark_as_advanced` still necessary?

msteinbeck replied:
Looks like @Spacelm [answers](https://github.com/conan-io/conan-center-index/pull/3149) this:

`add ALIAS targets tinyspline::tinyspline and tinysplinecxx::tinysplinecxx (consumers can use those targets, no matter if they rely on add_subdirectory or find_package)`

msteinbeck commented on src/tinyspline.h:
```diff
@@ -5,6 +5,22 @@
 #include <stddef.h>
 
+#ifdef _WIN32
+# define TINYSPLINE_SHARED_EXPORT __declspec(dllexport)
```
Usually, I do not indent definitions in conditions. Is this convention?

madebr replied:
You're the owner of tinyspline, so you make the rules.
I just thought it was more readable.
But I'll change it if you want.

msteinbeck replied:
I prefer consistency :). Thanks.

msteinbeck commented on src/tinyspline.h:
```diff
@@ -722,11 +738,11 @@ size_t ts_deboornet_multiplicity(const tsDeBoorNet *net);
 /**
  * Returns the dimension of \p net. The dimension of a net describes the number
- * of components for each point in ts_bspline_get_points(spline).
+ * of components for each point in TINYSPLINE_API ts_bspline_get_points(spline).
```
search and replace fail? :)

madebr replied:
You caught me red handed. 😄 
I did `s/ (ts_[a-z]+)/TINYSPLINE_API \1/`

msteinbeck commented on src/tinysplinecxx.h:
```diff
@@ -4,6 +4,14 @@
 #include <vector>
 #include <string>
 
+#ifdef TINYSPLINECXX_STATIC
```
__declspec(dllexport) is not required here?

madebr replied:
No, this allows using tinyspline inside a shared library without exporting the symbols.

msteinbeck replied:
I see.

msteinbeck commented on test/c/CMakeLists.txt:
```diff
@@ -13,8 +13,9 @@ file(GLOB_RECURSE TINYSPLINE_TESTS_SOURCE_FILES "*.c")
-	${TINYSPLINE_C_STATIC_LIBRARIES})
+	PRIVATE tinyspline)
+
+add_test(tinyspline_tests tinyspline_tests)
```
What is the purpose of this line?

madebr replied:
This change + `enable_testing()` in the root of the project allows executing `make test`.
It's the cmake interface to `ctest`.

msteinbeck replied:
Nice. Thanks.

madebr commented:
I just note that you have 2 calls to `project(tinyspline)`.
One in the main `CMakeLists.txt` and one in `src/CMakeLists.txt`.

The second one is superfluous. Is it ok if I remove it?

msteinbeck commented:
> Another possibility would be to add a option to enable -Werror/-WX.
Something like the following:

> option(TINYSPLINE_ENABLE_ERRORS "Enable -Werror/-WX" ON)

> This allows overriding, even when building as a subproject and does not require a compare in the main cmake script.

I like this idea. Yet, I would rename the variable to something like `TINYSPLINE_WARNINGS_AS_ERRORS`. I assume that no one would like to enable errors :D.

madebr commented:
> I like this idea. Yet, I would rename the variable to something like `TINYSPLINE_WARNINGS_AS_ERRORS`. I assume that no one would like to enable errors :D.

Haha, I was thinking the same thing while I was writing it.

Btw, this will only treat warnings as errors when compiling the library.
The tests + examples do not receive this treat.

msteinbeck commented:
> I just note that you have 2 calls to project(tinyspline).
> One in the main CMakeLists.txt and one in src/CMakeLists.txt.

> The second one is superfluous. Is it ok if I remove it?
> After this removal, running cmake src won't work anymore.
> You will have to use the main cmake script.

I like the fact that `src` can be used as a standalone project. This is very convenient when creating source packages with only the core libraries. That said, if you need to remove the superfluous `project(tinyspline)` call, go ahead.

madebr commented:
> 
> 
> > I just note that you have 2 calls to project(tinyspline).
> > One in the main CMakeLists.txt and one in src/CMakeLists.txt.
> 
> > The second one is superfluous. Is it ok if I remove it?
> > After this removal, running cmake src won't work anymore.
> > You will have to use the main cmake script.
> 
> I like the fact that `src` can be used as a standalone project. This is very convenient when creating source packages with only the core libraries. That said, if you need to remove the superfluous `project(tinyspline)` call, go ahead.

I've left it in, cmake is not complaining.
It's fine to have multiple calls to project, it's just weird to use the same name.
You can rename the main project to e.g. `tinyspline_main` or something else.

CMake creates some variables, such as `tinyspline_SOURCE_DIR`, `tinyspline_VERSION`, `tinyspline_XXX` which are now unknown/dependent of cmake's implementation.
It's fine as long as you're not using them.

msteinbeck commented on src/CMakeLists.txt:
```diff
@@ -11,7 +11,7 @@ set_property(GLOBAL PROPERTY USE_FOLDERS ON)
 list(APPEND CMAKE_MODULE_PATH "${CMAKE_CURRENT_SOURCE_DIR}/cmake")
 
-
+option(TINYSPLINE_WARNINGS_AS_ERRORS "Treat warnings as errors" ON)
```
There is a dedicated options section: https://github.com/msteinbeck/tinyspline/blob/master/src/CMakeLists.txt#L280. Please put this statement into this section.

madebr commented:
After this change, the shared libraries are not built/tested.
Before, the shared libraries were built, but not tested.

msteinbeck commented on src/CMakeLists.txt:
```diff
@@ -731,43 +748,43 @@ write_basic_package_version_file(
+	target_include_directories(tinysplinecxx
+			PUBLIC
+				$<BUILD_INTERFACE:${TINYSPLINE_CXX_INCLUDE_DIR}>
+				$<INSTALL_INTERFACE:${TINYSPLICE_INSTALL_INCLUDE_DIR}>)
```
Another typo (TINYSPLICE).

msteinbeck commented on CMakeLists.txt:
```diff
@@ -4,8 +4,19 @@ project(tinyspline)
-add_subdirectory(test)
-add_subdirectory(docs)
+if(TINYSPLINE_BUILD_EXAMPLES)
+    add_subdirectory(examples)
```
spaces -> tabs

msteinbeck commented on CMakeLists.txt:
```diff
@@ -4,8 +4,19 @@ project(tinyspline)
+    add_subdirectory(test)
+endif()
+if(TINYSPLINE_BUILD_DOCS)
+    add_subdirectory(docs)
```
spaces -> tabs

msteinbeck commented on src/CMakeLists.txt:
```diff
@@ -326,12 +327,16 @@ if(TINYSPLINE_INSTALL_LIBRARY_DIR STREQUAL "")
 
 option(TINYSPLINE_FLOAT_PRECISION "Build TinySpline with float precision." OFF)
 
+option(TINYSPLINE_WARNINGS_AS_ERRORS "Treat warnings as errors" ON)
```
All other options were documented above. One may find this superfluous, yet, it allows to add some more internal documentation. In this case you could list the compiler flags enabled with this option.

madebr replied:
I forgot to add it to the documentation.
It's good to keep it up to date.

msteinbeck commented:
After merging this PR I will run a last test and create a new tag (0.3.0). Ok?

msteinbeck commented:
This PR looks good to me. thank you very much for your effort.

madebr commented:
> After merging this PR I will run a last test and create a new tag (0.3.0). Ok?

Sounds good to me, but I would wait a few days so we (conan-center-index) can let the ci run on the pre-release. So we don't have to add any extra patches or you who need to create a patch release.

msteinbeck commented:
> Sounds good to me, but I would wait a few days so we (conan-center-index) can let the ci run on the pre-release. So we don't have to add any extra patches or you won't need to create a patch release.

Let me know when you are ready. TinySpline's current master is stable---that is, there are no known bugs.

madebr commented:
Hey @msteinbeck 
You can go go ahead with a release, as you please.
We'll rebase the conan package on the new release.
Thanks!

msteinbeck commented:
Hi @madebr,

I did some testing and found some issues regarding the export of definitions. They should be fixed now. Before declaring master stable, I have two more questions:

1. I wondering why `TINYSPLINE_SHARED_EXPORT ` and `TINYSPLINE_SHARED_IMPORT ` are both defined as `__declspec(dllexport)` ([source](https://github.com/msteinbeck/tinyspline/blob/master/src/tinyspline.h#L9)). I suspect one of the definitions is incorrect.

2. Shouldn't [TINYSPLINECXX_STATIC](https://github.com/msteinbeck/tinyspline/blob/master/src/tinysplinecxx.h#L7) imply [TINYSPLINE_STATIC](https://github.com/msteinbeck/tinyspline/blob/master/src/tinyspline.h#L16)? Currently, the C++ interface is build without `TINYSPLINE_STATIC` in any case.

Also, I will try to add some tests regarding the import of CMake targets.

madebr commented:
>     1. I wondering why `TINYSPLINE_SHARED_EXPORT ` and `TINYSPLINE_SHARED_IMPORT ` are both defined as `__declspec(dllexport)` ([source](https://github.com/msteinbeck/tinyspline/blob/master/src/tinyspline.h#L9)). I suspect one of the definitions is incorrect.

Good catch!
```
#define TINYSPLINE_SHARED_IMPORT __declspec(dllexport)
```
should be replaced with
```
#define TINYSPLINE_SHARED_IMPORT __declspec(dllimport)
```
I think this typo would re-export the symbols.

>     2. Shouldn't [TINYSPLINECXX_STATIC](https://github.com/msteinbeck/tinyspline/blob/master/src/tinysplinecxx.h#L7) imply [TINYSPLINE_STATIC](https://github.com/msteinbeck/tinyspline/blob/master/src/tinyspline.h#L16)? Currently, the C++ interface is build without `TINYSPLINE_STATIC` in any case.

Oh, another error!
While it it theoretically possible to have `tinyspline` and `tinysplinecxx` built as static and shared library, in reality they will always be built as the same type.
If you want to allow them having different type, then you should add `TINYSPLINECXX_STATIC` to the compile definitions of tinysplinecxx.
The easiest solution would be to just forget about `TINYSPLINECXX_STATIC` and rename it in `tinysplinecxx.h` to `TINYSPLINE_STATIC`.


> Also, I will try to add some tests regarding the import of CMake targets.

That would be great!
I have applied my suggestions above to current master, and pasted a patch below.
After this change, this test can re-use the `examples/c`, `examples/cxx` or `test/c` cmake script, because the aliased target is the same as the imported one.

<details>

<summary>cmake patch</summary>

```patch
diff --git a/examples/c/CMakeLists.txt b/examples/c/CMakeLists.txt
index ad08f68..3581e40 100644
--- a/examples/c/CMakeLists.txt
+++ b/examples/c/CMakeLists.txt
@@ -46,67 +46,67 @@ endif()
 ###############################################################################
 add_executable(quickstart_c quickstart.c)
 target_link_libraries(quickstart_c
-		PRIVATE tinyspline)
+		PRIVATE tinyspline::tinyspline)
 set_target_properties(quickstart_c PROPERTIES FOLDER "examples/c")
 
 add_executable(json_export_c json_export.c)
 target_link_libraries(json_export_c
-		PRIVATE tinyspline)
+		PRIVATE tinyspline::tinyspline)
 set_target_properties(json_export_c PROPERTIES FOLDER "examples/c")
 
 if("${TINYSPLINE_DEFINITIONS}" MATCHES "TINYSPLINE_FLOAT_PRECISION"
 		AND GLUT_FOUND AND OPENGL_FOUND)
 	add_executable(bspline bspline.c)
 	target_link_libraries(bspline
-		PRIVATE tinyspline
+		PRIVATE tinyspline::tinyspline
 		${OPENGL_LIBRARIES} ${GLUT_LIBRARIES})
 	set_target_properties(bspline PROPERTIES FOLDER "examples/c")
 
 	add_executable(nurbs nurbs.c)
 	target_link_libraries(nurbs
-		PRIVATE tinyspline
+		PRIVATE tinyspline::tinyspline
 		${OPENGL_LIBRARIES} ${GLUT_LIBRARIES})
 	set_target_properties(nurbs PROPERTIES FOLDER "examples/c")
 
 	add_executable(split split.c)
 	target_link_libraries(split
-		PRIVATE tinyspline
+		PRIVATE tinyspline::tinyspline
 		${OPENGL_LIBRARIES} ${GLUT_LIBRARIES})
 	set_target_properties(split PROPERTIES FOLDER "examples/c")
 
 	add_executable(beziers beziers.c)
 	target_link_libraries(beziers
-		PRIVATE tinyspline
+		PRIVATE tinyspline::tinyspline
 		${OPENGL_LIBRARIES} ${GLUT_LIBRARIES})
 	set_target_properties(beziers PROPERTIES FOLDER "examples/c")
 
 	add_executable(interpolation interpolation.c)
 	target_link_libraries(interpolation
-		PRIVATE tinyspline
+		PRIVATE tinyspline::tinyspline
 		${OPENGL_LIBRARIES} ${GLUT_LIBRARIES})
 	set_target_properties(interpolation PROPERTIES FOLDER "examples/c")
 
 	add_executable(derivative derivative.c)
 	target_link_libraries(derivative
-		PRIVATE tinyspline
+		PRIVATE tinyspline::tinyspline
 		${OPENGL_LIBRARIES} ${GLUT_LIBRARIES})
 	set_target_properties(derivative PROPERTIES FOLDER "examples/c")
 
 	add_executable(tension tension.c)
 	target_link_libraries(tension
-		PRIVATE tinyspline
+		PRIVATE tinyspline::tinyspline
 		${OPENGL_LIBRARIES} ${GLUT_LIBRARIES})
 	set_target_properties(tension PROPERTIES FOLDER "examples/c")
 
 	add_executable(triangle triangle.c)
 	target_link_libraries(triangle
-		PRIVATE tinyspline
+		PRIVATE tinyspline::tinyspline
 		${OPENGL_LIBRARIES} ${GLUT_LIBRARIES})
 	set_target_properties(triangle PROPERTIES FOLDER "examples/c")
 
 	add_executable(triangle-nurbs triangle-nurbs.c)
 	target_link_libraries(triangle-nurbs
-		PRIVATE tinyspline
+		PRIVATE tinyspline::tinyspline
 		${OPENGL_LIBRARIES} ${GLUT_LIBRARIES})
 	set_target_properties(triangle-nurbs PROPERTIES FOLDER "examples/c")
 endif()
diff --git a/examples/cxx/CMakeLists.txt b/examples/cxx/CMakeLists.txt
index e823ee3..7a49665 100644
--- a/examples/cxx/CMakeLists.txt
+++ b/examples/cxx/CMakeLists.txt
@@ -3,10 +3,10 @@
 ###############################################################################
 add_executable(quickstart quickstart.cpp)
 target_link_libraries(quickstart
-        PRIVATE tinysplinecxx)
+        PRIVATE tinyspline::tinysplinecxx)
 set_target_properties(quickstart PROPERTIES FOLDER "examples/cpp")
 
 add_executable(time_series time_series.cpp)
 target_link_libraries(time_series
-        PRIVATE tinysplinecxx)
+        PRIVATE tinyspline::tinysplinecxx)
 set_target_properties(time_series PROPERTIES FOLDER "examples/cpp")
diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index 8f5d87b..ec3fa94 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -735,6 +735,7 @@ list(JOIN TINYSPLINE_C_INSTALL_CMAKE_CONFIG_DIR "/"
 	TINYSPLINE_C_INSTALL_CMAKE_CONFIG_DIR)
 
 add_library(tinyspline ${TINYSPLINE_C_SOURCE_FILES})
+add_library(tinyspline::tinyspline ALIAS tinyspline)
 target_compile_definitions(tinyspline
 	PUBLIC ${TINYSPLINE_DEFINITIONS})
 set_target_properties(tinyspline PROPERTIES
@@ -796,6 +797,7 @@ if(TINYSPLINE_ENABLE_CXX)
 		TINYSPLINE_CXX_INSTALL_CMAKE_CONFIG_DIR)
 
 	add_library(tinysplinecxx ${TINYSPLINE_CXX_SOURCE_FILES})
+	add_library(tinyspline::tinysplinecxx ALIAS tinysplinecxx)
 	target_compile_definitions(tinysplinecxx
 		PUBLIC ${TINYSPLINE_DEFINITIONS})
 	set_target_properties(tinysplinecxx PROPERTIES
@@ -816,7 +818,7 @@ if(TINYSPLINE_ENABLE_CXX)
 		PUBLIC_HEADER DESTINATION ${TINYSPLINE_INSTALL_INCLUDE_DIR})
 	install(EXPORT tinysplinecxx
 		DESTINATION "${TINYSPLINE_CXX_INSTALL_CMAKE_CONFIG_DIR}"
-		NAMESPACE "${TINYSPLINE_CXX_LIBRARY_OUTPUT_NAME}::"
+		NAMESPACE "${TINYSPLINE_C_LIBRARY_OUTPUT_NAME}::"
 		FILE "${TINYSPLINE_CXX_LIBRARY_OUTPUT_NAME}-targets.cmake")
 	configure_package_config_file(
 		"${CMAKE_CURRENT_SOURCE_DIR}/pkg/tinysplinecxx-config.cmake.in"
diff --git a/src/pkg/tinysplinecxx.pc.in b/src/pkg/tinysplinecxx.pc.in
index 8dd64f3..2084482 100644
--- a/src/pkg/tinysplinecxx.pc.in
+++ b/src/pkg/tinysplinecxx.pc.in
@@ -6,6 +6,7 @@ includedir=${prefix}/@TINYSPLINE_INSTALL_INCLUDE_DIR@
 Name: @TINYSPLINE_PACKAGE_NAME@
 Description: @TINYSPLINE_SUMMARY@
 Version: @TINYSPLINE_VERSION@
+Requires: tinyspline=@TINYSPLINE_VERSION@
 
 Libs: -L${libdir} -l@TINYSPLINE_PKGCONFIG_CXX_LINK_LIBRARIES@
 Cflags: -I${includedir} @TINYSPLINE_PKGCONFIG_CXX_FLAGS@
diff --git a/src/tinysplinecxx.h b/src/tinysplinecxx.h
index a50c313..a18935f 100644
--- a/src/tinysplinecxx.h
+++ b/src/tinysplinecxx.h
@@ -4,7 +4,7 @@
 #include <vector>
 #include <string>
 
-#ifdef TINYSPLINECXX_STATIC
+#ifdef TINYSPLINE_STATIC
 #define TINYSPLINECXX_API
 #elif defined(tinysplinecxx_EXPORTS)
 #define TINYSPLINECXX_API TINYSPLINE_SHARED_EXPORT
diff --git a/test/c/CMakeLists.txt b/test/c/CMakeLists.txt
index 4fbfb06..ee56866 100644
--- a/test/c/CMakeLists.txt
+++ b/test/c/CMakeLists.txt
@@ -13,7 +13,7 @@ file(GLOB_RECURSE TINYSPLINE_TESTS_SOURCE_FILES "*.c")
 include_directories(cutest)
 add_executable(tinyspline_tests ${TINYSPLINE_TESTS_SOURCE_FILES})
 target_link_libraries(tinyspline_tests
-	PRIVATE tinyspline)
+	PRIVATE tinyspline::tinyspline)
 
 add_test(tinyspline_tests tinyspline_tests)
 
```
</details>

msteinbeck commented:
> While it it theoretically possible to have tinyspline and tinysplinecxx built as static and shared library respectively,
in reality they will always be built as the same type.
On the other side, allowing them to have different type, then it would be possible to create a single shared library containing both tinyspline and tinysplinecxx.
If you want to allow them having different type, then you should add TINYSPLINECXX_STATIC to the compile definitions of tinysplinecxx.
The easiest solution would be to just forget about TINYSPLINECXX_STATIC and rename it in tinysplinecxx.h to TINYSPLINE_STATIC.

Following your advice, `tinyspline` and `tinysplinecxx` have different defines (`TINYSPLINE_STATIC` and `TINYSPLINECXX_STATIC` respectively). When building the C++ interface `TINYSPLINE_STATIC` is also defined ([link](https://github.com/msteinbeck/tinyspline/blob/master/src/CMakeLists.txt#L492)).

> That would be great!
I have applied my suggestions above to current master, and pasted a patch below.
After this change, this test can re-use the examples/c, examples/cxx or test/c cmake script, because the aliased target is the same as the imported one.

I added a simple test project ([link](https://github.com/msteinbeck/tinyspline/tree/master/test/pkg)). The project contains a [CMakeLists.txt](https://github.com/msteinbeck/tinyspline/blob/master/test/pkg/CMakeLists.txt) and a [Makefile](https://github.com/msteinbeck/tinyspline/blob/master/test/pkg/Makefile). While the `CMakeLists.txt`---obviously---tests the import and linkage of `tinyspline` and `tinysplinecxx`, the `Makefile` links against `tinyspline` and `tinysplinecxx` using pkg-config. The source files compiled by `CMakeLists.txt` and `Makefile` ([test.c](https://github.com/msteinbeck/tinyspline/blob/master/test/pkg/test.c) and [test.cxx](https://github.com/msteinbeck/tinyspline/blob/master/test/pkg/test.cxx)), in turn, contain some simple tests using the `assert()` macro.

Actually, I found a small typo in [tinysplinecxx-config.cmake.in](https://github.com/msteinbeck/tinyspline/pull/167/files#diff-e9a3006d446703d71761eb2b40b017f2cf76d1d002f3fddb25e52638e1d09fceR8). `TINYSPLINE_LIBRARIES` must be replaced with `TINYSPLINECXX_LIBRARIES`.

Now I need to run those tests in CI. While this should be easy on Linux and OSX, I need to figure out where files are installed with `cmake --build . --target install` on Windows.

madebr commented:
> Following your advice, `tinyspline` and `tinysplinecxx` have different defines (`TINYSPLINE_STATIC` and `TINYSPLINECXX_STATIC` respectively). When building the C++ interface `TINYSPLINE_STATIC` is also defined ([link](https://github.com/msteinbeck/tinyspline/blob/master/src/CMakeLists.txt#L492)).

I'll leave it up to you whether you want to keep `TINYSPLINECXX_STATIC` or not.


> > That would be great!
> > I have applied my suggestions above to current master, and pasted a patch below.
> > After this change, this test can re-use the examples/c, examples/cxx or test/c cmake script, because the aliased target is the same as the imported one.
> 
> I added a simple test project ([link](https://github.com/msteinbeck/tinyspline/tree/master/test/pkg)). The project contains a [CMakeLists.txt](https://github.com/msteinbeck/tinyspline/blob/master/test/pkg/CMakeLists.txt) and a [Makefile](https://github.com/msteinbeck/tinyspline/blob/master/test/pkg/Makefile). While the `CMakeLists.txt`---obviously---tests the import and linkage of `tinyspline` and `tinysplinecxx`, the `Makefile` links against `tinyspline` and `tinysplinecxx` using pkg-config. The source files compiled by `CMakeLists.txt` and `Makefile` ([test.c](https://github.com/msteinbeck/tinyspline/blob/master/test/pkg/test.c) and [test.cxx](https://github.com/msteinbeck/tinyspline/blob/master/test/pkg/test.cxx)), in turn, contain some simple tests using the `assert()` macro.

The tests look great, I would change the following:
**CMakeLists.txt
- Fail when not finding the dependencies:
   ```cmake
   find_package(tinyspline REQUIRED)
   find_package(tinysplinecxx REQUIRED)
   ```
- add the following tests:
    ```cmake
    if(NOT TARGET tinespline::tinyspline)
        message(FATAL_ERROR "tinyspline::tinyspline target not defined")
    endif()
    if(NOT TARGET tinesplinecxx::tinysplinecxx)
        message(FATAL_ERROR "tinysplinecxx::tinysplinecxx target not defined")
    endif()
    ```
- Consider using `SEND_ERROR` instead of `FATAL_ERROR`.
Then the cmake script will continue running but will fail at the end.


> Actually, I found a small typo in [tinysplinecxx-config.cmake.in](https://github.com/msteinbeck/tinyspline/pull/167/files#diff-e9a3006d446703d71761eb2b40b017f2cf76d1d002f3fddb25e52638e1d09fceR8). `TINYSPLINE_LIBRARIES` must be replaced with `TINYSPLINECXX_LIBRARIES`.

Great bugfinding. I didn't check the variables as targets are easier to handle.

> Now I need to run those tests in CI. While this should be easy on Linux and OSX, I need to figure out where files are installed with `cmake --build . --target install` on Windows.

The installation prefix of tinyspline on Windows looks like the following:
```
├── bin
│   ├── tinyspline.dll
│   └── tinysplinecxx.dll
├── include
│   ├── tinyspline.h
│   └── tinysplinecxx.h
└── lib64
    ├── cmake
    │   ├── tinyspline
    │   │   ├── tinyspline-config.cmake
    │   │   ├── tinyspline-config-version.cmake
    │   │   ├── tinyspline-targets.cmake
    │   │   └── tinyspline-targets-debug.cmake
    │   └── tinysplinecxx
    │       ├── tinysplinecxx-config.cmake
    │       ├── tinysplinecxx-config-version.cmake
    │       ├── tinysplinecxx-targets.cmake
    │       └── tinysplinecxx-targets-debug.cmake
    ├── pkgconfig
    │   ├── tinyspline.pc
    │   └── tinysplinecxx.pc
    ├── tinyspline.lib
    └── tinysplinecxx.lib
```

madebr commented:
The current cmake script does not support parallel installs of a static/shared tinyspline, do you want that?
Also, the current cmake script does also not support parallel installs of debug/release/minsizerel install, do you want that?

msteinbeck commented:
> The current cmake script does not support parallel installs of a static/shared tinyspline on MSVC, do you want that?
Also, the current cmake script does also not support parallel installs of debug/release/minsizerel install, do you want that?

I'm afraid that adding support for parallel installs of static/shared libs as well as of debug/release/minsizerel builds makes the CMake script (in particular the generation of the CMake config and pkg-config files) rather complex. Or am I misjudging the effort here?

Anyhow, libraries and config scripts are already generated and installed based on [TINYSPLINE_C_LIBRARY_OUTPUT_NAME](https://github.com/msteinbeck/tinyspline/blob/master/src/CMakeLists.txt#L270) and [TINYSPLINE_CXX_LIBRARY_OUTPUT_NAME](https://github.com/msteinbeck/tinyspline/blob/master/src/CMakeLists.txt#L273). That is, one could change these values (e.g., to `tinysplinecxx_dbg`) and install multiple configurations of the libraries side-by-side without conflicts. The desired configuration can then be imported by passing the corresponding name to `find_package` (e.g., `find_package(tinysplinecxx_dbg)`). For the sake of convenience, `TINYSPLINE_C_LIBRARY_OUTPUT_NAME` and `TINYSPLINE_CXX_LIBRARY_OUTPUT_NAME` (at least implicitly) must be exposed as options though.

madebr commented:
> 
> 
> > The current cmake script does not support parallel installs of a static/shared tinyspline on MSVC, do you want that?
> > Also, the current cmake script does also not support parallel installs of debug/release/minsizerel install, do you want that?
> 
> I'm afraid that adding support for parallel installs of static/shared libs as well as of debug/release/minsizerel builds makes the CMake script (in particular the generation of the CMake config and pkg-config files) rather complex. Or am I misjudging the effort here?

Parallel installs of shared/static or debug/release/minsizerel can be done by giving these different postfixes.
e.g.:
- `tinyspline_static` for static msvc
- `tinyspline` for shared msvc + static non-msvc + shared non-msvc
- `tinyspline_debug` for shared debug
- `tinyspline_static_minsizerel` for static minsizerel

The cmake script should just be careful to not overwrite libraries of another configuration.

> Anyhow, libraries and config scripts are already generated and installed based on [TINYSPLINE_C_LIBRARY_OUTPUT_NAME](https://github.com/msteinbeck/tinyspline/blob/master/src/CMakeLists.txt#L270) and [TINYSPLINE_CXX_LIBRARY_OUTPUT_NAME](https://github.com/msteinbeck/tinyspline/blob/master/src/CMakeLists.txt#L273). That is, one could change these values (e.g., to `tinysplinecxx_dbg`) and install multiple configurations of the libraries side-by-side without conflicts. The desired configuration can then be imported by passing the corresponding name to `find_package` (e.g., `find_package(tinysplinecxx_dbg)`). For the sake of convenience, `TINYSPLINE_C_LIBRARY_OUTPUT_NAME` and `TINYSPLINE_CXX_LIBRARY_OUTPUT_NAME` (at least implicitly) must be exposed as options though.

I didn't know of this tinyspline option. I was thinking of using [`CMAKE_<CONFIG>_POSTFIX`](https://cmake.org/cmake/help/latest/variable/CMAKE_CONFIG_POSTFIX.html). But if it isn't required or the complexity is too high or the user can get confused, maybe it's better to avoid it.

msteinbeck commented:
> 
    Fail when not finding the dependencies:

    find_package(tinyspline REQUIRED)
    find_package(tinysplinecxx REQUIRED)

    add the following tests:

    if(NOT TARGET tinespline::tinyspline)
        message(FATAL_ERROR "tinyspline::tinyspline target not defined")
    endif()
    if(NOT TARGET tinesplinecxx::tinysplinecxx)
        message(FATAL_ERROR "tinysplinecxx::tinysplinecxx target not defined")
    endif()

    Consider using SEND_ERROR instead of FATAL_ERROR.
    Then the cmake script will continue running but will fail at the end.

Good idea. Done.

msteinbeck commented:
> I didn't know of this tinyspline option. I was thinking of using CMAKE_<CONFIG>_POSTFIX. But if it isn't required or the complexity is too high or the user can get confused, maybe it's better to avoid it.

I'll have a look into it. Integrating `CMAKE_<CONFIG>_POSTFIX` into the current approach should be possible. Instead of a `TINYSPLINE_...` option, one would use  `CMAKE_<CONFIG>_POSTFIX` to configure the names of the libraries and config scripts etc. Thanks for the hints!

madebr commented:
I think there is another issue if you want to support parallel installs of shared and static libraries.
The generated `tinyspline-targets.cmake` file (on Linux) contains the lines
```
add_library(tinyspline::tinyspline STATIC IMPORTED)

set_target_properties(tinyspline::tinyspline PROPERTIES
  INTERFACE_COMPILE_DEFINITIONS "TINYSPLINE_STATIC"
  INTERFACE_INCLUDE_DIRECTORIES "${_IMPORT_PREFIX}/include"
  INTERFACE_LINK_LIBRARIES "-lm"
)
```
This will clash with a shared library because of the library type and the compile definitions.
This can be fixed by being able to build both libraries, as was possible before.
Using some functions, this can be added .

```cmake
# src/CMakeLists.txt: generate both shared/static libraries

option(TINYSPLINE_SHARED "Build shared tinyspline libraries" ${BUILD_SHARED_LIBS})

set(_tinespline_static_default ON)
if(BUILD_SHARED_LIBS)
    set(_tinespline_static_default OFF)
endif()

option(TINYSPLINE_STATIC "Build static tinyspline libraries" ${_tinespline_static_default})

if(NOT TINYSPLINE_SHARED AND NOT TINYSPLINE_STATIC)
    message(FATAL_ERROR "TINYSPLINE_SHARED and TINYSPLINE_STATIC cannot be both disabled")
endif()

function(add_tinyspline_library NAME TYPE)
    # Create tinyspline library of unknown type + tinyspline::NAME alias
    add_library(${NAME} ${TYPE} ${SOURCES} ...)
    add_library(tinyspline::${NAME} ALIAS ${NAME})
    ...
    install(TARGETS ${NAME} EXPORT tinyspline::)
endif()

if(TINYSPLINE_SHARED)
    add_tinyspline_library(tinyspline STATIC)
endif()
if(TINYSPLINE_STATIC)
    add_tinyspline_library(tinyspline_static STATIC)
    target_compile_definitions(tinyspline_static PUBLIC TINYSPLINE_STATIC)
endif()

function(add_tinysplinecxx_library NAME TYPE)
    add_library(${NAME} ${TYPE} ${SOURCES} ...)
    add_library(tinysplinecxx::${NAME} ALIAS ${NAME})
    ...
    install(TARGETS ${NAME} EXPORT tinyspline::)
    ...
endif()

if(TINYSPLINE_SHARED)
    add_tinyspline_library(tinysplinecxx STATIC)
    target_link_libraries(tinysplinecxx PRIVATE tinyspline)
endif()
if(TINYSPLINE_STATIC)
    add_tinyspline_library(tinysplinecxx_static STATIC)
    target_compile_definitions(tinysplinecxx_static PUBLIC TINYSPLINECXX_STATIC)
    target_link_libraries(tinysplinecxx_static PRIVATE tinyspline_static)
endif()
```

The changes in `src/CMakeLists.txt` will raise the question on what library to test? The shared or static library? Or one? Which one?

```cmake
# test/c/CMakeLists.txt: possibility 1, link to both shared and static

add_executable(test_shared ...)
target_link_libraries(test_shared PRIVATE tinyspline::tinyspline)

add_executable(test_shared ...)
target_link_libraries(test_static PRIVATE tinyspline::tinyspline_static)

# test/c/CMakeLists.txt: possibility 2, link to only shared or static (chosen by option)

option(TINYSPLINE_TEST_SHARED "Test the shared tinyspline libraries" ${TINYSPLINE_SHARED})
if((TINYSPLINE_TEST_SHARED AND NOT TINYSPLINE_SHARED) OR (NOT TINYSPLINE_TEST_SHARED AND NOT TINYSPLINE_STATIC))
    message(FATAL_ERROR "TINYSPLINE_TEST_SHARED not set correctly")
endif()
if(TINYSPLINE_TEST_SHARED)
    add_library(tinyspline::tinyspline_test ALIAS tinyspline::tinyspline)
else()
    add_library(tinyspline::tinyspline_test ALIAS tinyspline::tinyspline_static)
endif()

add_executable(test_tinyspline ...)
target_link_libraries(test_static PRIVATE tinyspline::tinyspline_test
```

`tinyspline-config.cmake.in` and `tinysplinecxx-config.cmake.in` also need to be modified to supply because `tinyspline::tinyspline` and `tinysplinecxx::tinysplinecxx` are not always available...

Having both available does not look like such a trivial change after all...

msteinbeck commented:
I tested the compilation of the shared libraries under Windows with MSVC and got several errors with message `[...] inconsistent dll linkage [...]`. I found that `#elif defined(tinyspline_EXPORTS)` and `#elif defined(tinysplinecxx_EXPORTS)` must be explicitly defined with CMake---I thought MSVC defines them when building the libraries because your PR didn't add them to the definitions of the corresponding targets. Anyhow, I fixed the issues so that the shared libraries can now be built with and without CMake. Also, I replaced `TINYSPLINE_STATIC` with `TINYSPLINE_SHARED` /`TINYSPLINECXX_STATIC` with `TINYSPLINECXX_SHARED`:

```
#ifdef TINYSPLINE_SHARED
#ifdef TINYSPLINE_EXPORT
#define TINYSPLINE_API TINYSPLINE_SHARED_EXPORT
#else
#define TINYSPLINE_API TINYSPLINE_SHARED_IMPORT
#endif
#else
#define TINYSPLINE_API
#endif
```

The reason is that the CMake script, by default, builds the static version of the libraries. Thus, client code linking against the libraries (without CMake, pkg-config etc.) do not have to add additional defines in order to use the libraries. The same is true if the source code is simply included into another project.

msteinbeck commented:
> Having both available does not look like such a trivial change after all...

I'll leave out this feature for the upcoming 0.3.0 release. Let's not develop solutions for something that is not yet an issue.

msteinbeck commented:
I also replaced:

```
#ifdef _WIN32
...
#endif
```

with:

```
#ifdef _MSC_VER
...
#endif
```

as `__declspec(dllexport)` and `__declspec(dllimport)` are only necessary when compiling code with MSVC but not with, for example. MinGW.

madebr commented:
> I tested the compilation of the shared libraries under Windows with MSVC and got several errors with message `[...] inconsistent dll linkage [...]`. I found that `#elif defined(tinyspline_EXPORTS)` and `#elif defined(tinysplinecxx_EXPORTS)` must be explicitly defined with CMake

The `_EXPORTS` must only be defined when compiling the library, not when using the library.

> ---I thought MSVC defines them when building the libraries because your PR didn't add them to the definitions of the corresponding targets.

CMake automatically defines `<TARGET>_EXPORTS` when `TARGET` is built as a shared library.
The name of this define can be modified by setting the [`DEFINE_SYMBOL`](https://cmake.org/cmake/help/latest/prop_tgt/DEFINE_SYMBOL.html) property.

> Anyhow, I fixed the issues so that the shared libraries can now be built with and without CMake. Also, I replaced `TINYSPLINE_STATIC` with `TINYSPLINE_SHARED` /`TINYSPLINECXX_STATIC` with `TINYSPLINECXX_SHARED`:
> 
> ```
> #ifdef TINYSPLINE_SHARED
> #ifdef TINYSPLINE_EXPORT
> #define TINYSPLINE_API TINYSPLINE_SHARED_EXPORT
> #else
> #define TINYSPLINE_API TINYSPLINE_SHARED_IMPORT
> #endif
> #else
> #define TINYSPLINE_API
> #endif
> ```
> 
> The reason is that the CMake script, by default, builds the static version of the libraries. Thus, client code linking against the libraries (without CMake, pkg-config etc.) do not have to add additional defines in order to use the libraries. The same is true if the source code is simply included into another project.

If I read this code correctly, using cmake you can do:
```cmake
set_target_properties(tinyspline PROPERTIES DEFINE_SYMBOL TINYSPLINE_EXPORT)
```
Other build script should also define this symbol when creating a shared library.

> I also replaced:
> 
> ```
> #ifdef _WIN32
> ```
> with:
> ```
> #ifdef _MSC_VER
> ```
> as `__declspec(dllexport)` and `__declspec(dllimport)` are only necessary when compiling code with MSVC but not with, for example. MinGW.

This is ok.
Keeping it, and defining `TINYSPLINE_EXPORT` to `__attribute__((visibility("default"))` (for non-msvc compiler), would allow you to use `-fvisibility=hidden`. This will only make the api visible from outside the shared library, even on Linux.

madebr commented:
I noticed that the `BUILD_SHARED_LIBS=ON` + MSVC tests on travis are failing.
See https://travis-ci.org/github/msteinbeck/tinyspline/jobs/736703287#L332

I believe this is because the dll and exe are created in different directories, the exe will fail because the dll cannot be found.
On Linux, this does not fail because the test executable has rpath set (which is removed when installed).
I'm not sure, but may using `cmake --build --target test` will work.
This will use ctest to run all tests added with `add_test` (by this pr).
It also works from the root directory because of the `enable_testing` function call in the main cmake script.

msteinbeck commented:
> I'm not sure, but may using cmake --build --target test will work.
This will use ctest to run all tests added with add_test (by this pr).
It also works from the root directory because of the enable_testing function call in the main cmake script.

Running `cmake --build . --target test` fails on Windows (saying that test.vcxproj does not exist). However, I found that `ctest` can be run directly. Yet, this does not fix the dll issue. After some tinkering I found a suitable solution: CMake allows to set certain environment variables for tests. So I added the following:

```
if(${CMAKE_SYSTEM_NAME} STREQUAL "Windows")
	set_tests_properties(tinyspline_tests PROPERTIES
		ENVIRONMENT "PATH=${TINYSPLINE_OUTPUT_DIRECTORY};$ENV{PATH}")
endif()
```

madebr commented:
On MSVC, you can use the vcvars script + the cmake `NMake Makefiles` generator.
You can then use make/nmake like you would on Linux.

msteinbeck commented:
I am slowly making progress with travis. Only OSX with shared libraries and Linux with static library, clang, and pkg-config is still failing. I really hate fixing bugs related to travis...

I guess I can fix OSX by compiling the binary with rpath rather than setting `DYLD_LIBRARY_PATH` (it seems that setting `DYLD_LIBRARY_PATH` does not work in travis).

madebr commented:
> I am slowly making progress with travis. Only OSX with shared libraries and Linux with static library, clang, and pkg-config is still failing. I really hate fixing bugs related to travis...

I share the pain. I improved my *travis skills* by using docker (ubuntu) images.
As these are minimal, and repeatable, you can test your build script and copy the *golden* commands to the `.travis.yml` script.
I also do the travis work in some temporary git branch, not the main one, just to avoid clutter in the commit history.

> I guess I can fix OSX by compiling the binary with rpath rather than setting `DYLD_LIBRARY_PATH` (it seems that setting `DYLD_LIBRARY_PATH` does not work in travis).

On Macos, SIP exists. For the build use cases, it does not pass the `DYLD_LIBRARY_PATH` variable through.
So when you do e.g. `DYLD_LIBRARY_PATH=/x/y/z cmake --build . --target test`, the tested executables will not have a `DYLD_LIBRARY_PATH` variable set.
CMake has the [`ENVIRONMENT`](https://cmake.org/cmake/help/latest/prop_test/ENVIRONMENT.html) variable available. Maybe the following helps?

```cmake
if(APPLE)
    set_tests_properties(test PROPERTIES ENVIRONMENT DYLD_LIBRARY_PATH=$<$<TARGET_LINKER_FILE_DIR:tgt>:tinespline>)
elseif(WIN32)
    set_tests_properties(test PROPERTIES ENVIRONMENT PATH=$<$<TARGET_LINKER_FILE_DIR:tgt>:tinespline>)
else()
    set_tests_properties(test PROPERTIES ENVIRONMENT LD_LIBRARY_PATH=$<$<TARGET_LINKER_FILE_DIR:tgt>:tinespline>)
endif()
```
I haven't tried this before. So I can't guarantee success.

madebr commented:
I took a glance at the travis logs.

- Linux, static: https://travis-ci.org/github/msteinbeck/tinyspline/jobs/740698568
  I think the `relocation R_X86_64_32S against `.data' can not be used when making a PIE object; recompile with -fPIC` error is due to ASLR. See this [obligatory stackoverflow answer](https://stackoverflow.com/questions/2463150/what-is-the-fpie-option-for-position-independent-executables-in-gcc-and-ld/5030518#5030518). It's only clang that has this error btw.
- Macos, shared. I think this is causes by SIP. In addition to adding `DYLD_LIBRARY_PATH/LD_LIBRARY_PATH` in the Makefile/cmake script, you can also, while creating the test executable, pass [`-Wl,-rpath=/path/to/lib/dir`](https://ftp.gnu.org/old-gnu/Manuals/ld-2.9.1/html_node/ld_3.html) to the compiler.

msteinbeck commented:
I replaced TravisCI with Github actions (which is way faster). I still need to adapt CI for cross-compiling though. However, the new release should be ready soon.

msteinbeck commented:
TinySpline has now a new release: 0.3.0 :).

madebr commented:
@msteinbeck 
Great!

I just updated the tinyspline recipe in conan-center-index: see https://github.com/conan-io/conan-center-index/pull/3639

I had to apply a patch that replaces `CMAKE_BINARY_DIR` with `CMAKE_CURRENT_BINARY_DIR`.
We build tinyspline as a subproject, which causes `CMAKE_BINARY_DIR` != `CMAKE_CURRENT_BINARY_DIR`.

msteinbeck commented:
> I had to apply a patch that replaces CMAKE_BINARY_DIR with CMAKE_CURRENT_BINARY_DIR.
We build tinyspline as a subproject, which causes CMAKE_BINARY_DIR != CMAKE_CURRENT_BINARY_DIR.

I will fix that for the next release.
